src/client_side.cc

```diff
@@ -2405,16 +2420,23 @@ tlsAttemptHandshake(ConnStateData *conn, PF *callback)
             debugs(83, (xerrno == ECONNRESET) ? 1 : 2, "Error negotiating SSL connection on FD " << fd << ": " <<
                    (xerrno == 0 ? Security::ErrorString(ssl_error) : xstrerr(xerrno)));
         }
+        // XXX: The client may not have actually closed the connection.
+        // TODO: Detail ssl_error (via Ssl::ErrorDetail?) when appropriate.
+        // TODO: Detail system error (xerrno) when appropriate.
```

Comment:
Similar to tunnelOnError() discussed in another change request, I suspect that tlsAttemptHandshake() should detail its own errors (before returning a negative result). Most likely, you will want to detail (both?) `ssl_error` and `xerrno` errors in a new separate method, but other cases can probably be done inline.

If you are not sure how to proceed, then remove the `errDetail` parameter and then try to _start_ untangling this by handling non-zero xerrno values first (because they are probably easier to detail) while keeping future `ssl_error` needs in mind. What it would take to detail the _system_ error inplace (instead of returning the too-general and often lying `ERR_DETAIL_TLS_CLIENT_CLOSED` to the caller)?

Reply:
Still I am not able to decide what to do here. I have the following problems here:
  - The Ssl::ErrorDetail is designed to report certificates validation problems. But ok can be modified to hold other errors to.
 - An error page which can hold an Ssl::ErrorDetail structure does not make sense (the client connection is closed)
 - We are closing the connection and we do not have where to report the ssl_error and xerrno. We can log only an error code and an error detail code.
- The ssl_error code number does not make any sense to us, they are integer values which can be interpreted only by openSSL.

Does it make sense to log error details such as ERR_DETAIL_XXX where XXX is xerrno code. For ssl_error codes something like "ERR_DETAIL_SSL(an ERR_error_string())" code?

The  ERR_error_string openSSL function return a string  in the form "error:[error code]:[library name]:[function name]:[reason string]". We can use the ERR_reason_error_string which returns only the reason string.

Reply:
> The Ssl::ErrorDetail is designed to report certificates validation problems.

Not really. Its design is suitable for all kinds of SSL-related problems. We even have a non-validation-spefic SQUID_ERR_SSL_HANDSHAKE error.

> An error page which can hold an Ssl::ErrorDetail structure does not make sense (the client connection is closed)

That is debatable. Even the last part (i.e., the closed client connection) may not be true for many handshake errors. If creating an error page simplifies things a lot for us, then we should create it, regardless of whether the from-client connection can deliver the error to the client or whether the current Squid code delivers plain-text errors to expected-to-talk-TLS clients.

> We are closing the connection and we do not have where to report the ssl_error and xerrno. We can log only an error code and an error detail code.

Logging is reporting.

> The ssl_error code number does not make any sense to us, they are integer values which can be interpreted only by openSSL.

We are reporting the error code to the admin. The admin (or their support team) can use OpenSSL tools like [errstr](https://www.openssl.org/docs/manmaster/man1/openssl-errstr.html) to convert the numeric code into a more informative string that they can, for example, google with.

> Does it make sense to log error details such as ERR_DETAIL_XXX where XXX is xerrno code. For ssl_error codes something like "ERR_DETAIL_SSL(an ERR_error_string())" code?

We already have errorDetailName() API for this. It is already used for system errors AFAICT. For built-in OpenSSL error codes, I would use an "OpenSSL" and, if appropriate, "GnuTLS" names. We will need to find a way to distinguish built-in OpenSSL error codes from other error details. I think it is time to convert err_detail_type into a class that holds the type and the value as two separate fields (instead of using ranges). We can still use old ranges for old values, of course. I hope this conversion can be done without changing most existing callers -- they can continue to use enum constants like ERR_DETAIL_CLT_REQMOD_ABORT.

Reply:
And, again, consider starting with handling system errors because they may be easier to handle than SSL errors. For example, they do not require enhancing err_detail_type.

Reply:
Now there is an ErrorDetail and kids class hierarchy to store various errors.
 
The Ssl::ErrorDetail now is an ErrorDetail kid class. It stores some squid custom SSL/TLS errors, openSSL certificate verify errors and openSSL library errors as hex numbers which can be examined with the "openssl errstr" utility.